### PR TITLE
fix issue #629 and  some cleanups

### PR DESCRIPTION
**Pull request by DreamWuGit**
fix issue #629 and add test for it.
cleanup trace test  `tracer_err_insufficient_balance` useless code and correct comment
cleanup unused `StepRow`

### CODE REVIEW

File: bus-mapping/src/circuit_input_builder/input_state_ref.rs [hunk 1]
```diff
@@ -839,7 +839,7 @@ impl<'a> CircuitInputStateRef<'a> {
         let call = self.call()?;
 
         // Return from a call with a failure
-        if step.depth != next_depth && next_result.is_zero() {
+        if step.depth == next_depth + 1 && next_result.is_zero() {
```
Comment (han0110) [inline bus-mapping/src/circuit_input_builder/input_state_ref.rs:842]: Would you like to also change line 904 to `step.depth == next_depth + 1`? It seems more strict rules first can help detect such kinds of issue.

File: bus-mapping/src/circuit_input_builder/tracer_tests.rs [hunk 2]
```diff
@@ -246,25 +246,75 @@ fn tracer_err_insufficient_balance() {
     let next_step = block.geth_traces[0].struct_logs.get(index + 1);
     assert_eq!(step.error, None);
     assert_eq!(next_step.unwrap().op, OpcodeId::PUSH2);
-    assert_eq!(next_step.unwrap().stack, Stack(vec![Word::zero()])); // success = 0
+    assert_eq!(next_step.unwrap().stack, Stack(vec![Word::zero()])); // failure = 0
```
Comment (han0110) [inline bus-mapping/src/circuit_input_builder/tracer_tests.rs:249]: hmm I think the value on stack means `success`.

File: bus-mapping/src/circuit_input_builder/tracer_tests.rs [hunk 3]
```diff
@@ -246,25 +246,75 @@ fn tracer_err_insufficient_balance() {
...
+    PUSH32(Word::from(0x1000)) // value
+    PUSH32(Word::from(0x000000000000000000000000000000000cafe001)) // addr
+    PUSH32(0x1_0000) // gas
+    CALL
+
+    PUSH2(0xaa)
+    };
+    let code_b = bytecode! {
+    STOP
+    };
```
Comment (han0110) [inline bus-mapping/src/circuit_input_builder/tracer_tests.rs:274]:
```suggestion
    let code_a = bytecode! {
        PUSH1(0x0) // retLength
        PUSH1(0x0) // retOffset
        PUSH1(0x0) // argsLength
        PUSH1(0x0) // argsOffset
        PUSH32(Word::from(0x1000)) // value
        PUSH32(Word::from(0x000000000000000000000000000000000cafe001)) // addr
        PUSH32(0x1_0000) // gas
        CALL

        PUSH2(0xaa)
    };
    let code_b = bytecode! {
        STOP
    };
```

**Review by han0110 — COMMENTED** [review]
Thanks for fixing this! Besides some suggestions, others look good to me.

**Review by ed255 — APPROVED** [review]
Thanks for the fix!  Looks good to me, modulo the comments made by Han.

File: bus-mapping/src/circuit_input_builder/input_state_ref.rs (hunk 1 above)
Comment (DreamWuGit) [reply bus-mapping/src/circuit_input_builder/input_state_ref.rs:842]: Good to try updating , thanks

Comment (DreamWuGit) [reply bus-mapping/src/circuit_input_builder/input_state_ref.rs:842]: updated in [bda71b7](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/630/commits/bda71b728670f0567cf5e8961af6590f7d6b372f)

File: bus-mapping/src/circuit_input_builder/tracer_tests.rs (hunk 3 above)
Comment (DreamWuGit) [reply bus-mapping/src/circuit_input_builder/tracer_tests.rs]: updated in [bda71b7](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/630/commits/bda71b728670f0567cf5e8961af6590f7d6b372f). thanks

File: bus-mapping/src/circuit_input_builder/tracer_tests.rs (hunk 2 above)
Comment (DreamWuGit) [reply bus-mapping/src/circuit_input_builder/tracer_tests.rs:249]: @han0110 I understand the stack value is for call result in https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L698. in this case call's `ret` is zero, right ?

Comment (han0110) [reply bus-mapping/src/circuit_input_builder/tracer_tests.rs:249]: Yes

**Review by han0110 — APPROVED** [review]

**Pull request merged**